### PR TITLE
At 60 implement editing student info on the frontend

### PR DESCRIPTION
**Pull request by RyanVanDijck**

### CODE REVIEW

File: src/components/Students/StudentInfo.test.js [hunk 1]
```diff
@@ -1,25 +1,93 @@
...
+    <StudentInfo />
+  )
+
+  const firstNameSubmitButton = await screen.findByText(/Change First Name/i)
+  fireEvent.click(firstNameSubmitButton)
+
+  const firstNameInput = await screen.findByPlaceholderText(/Mike/i)
+  userEvent.type(firstNameInput,'Jack')
+
+  global.fetch = jest.fn(() => Promise.reject())
```
[the remark this answers is not in the document]
Comment (RyanVanDijck) [reply src/components/Students/StudentInfo.test.js]: Changed here:
https://github.com/digital-futures-academy/atlas-frontend/pull/25/commits/ea38ea3acc8ccdd88c4bb87574696acd6915951b

File: src/components/Students/StudentInfo.js [hunk 2]
```diff
@@ -22,24 +24,89 @@ class StudentInfo extends Component{
...
+        }
+        else{
+            return null; 
+        }
+    }
+
+    mainPage = () => {
         return (
             <div>
+                <this.check />
```
Comment (dearshrewdwit) [inline src/components/Students/StudentInfo.js:45]: What's this doing?

File: src/components/Students/StudentInfo.js [hunk 3]
```diff
@@ -22,24 +24,89 @@ class StudentInfo extends Component{
         .then(res => res.json())
         .then(data =>  this.setState({avatar_Url: data.avatar_url}))
     }
-    render(){
+
+    check = () => {
```
Comment (dearshrewdwit) [inline src/components/Students/StudentInfo.js:28]: When does this fire?

Comment (dearshrewdwit) [reply src/components/Students/StudentInfo.js:28]: Instead - use the data from the response.

Your endpoint should be responsible for sending back a response with the information you need - for example on update:

Success:
```json
{
  "status": "success",
  "student": {
  },
  "errors": []
}
```
failure:
```json
{
  "status": "failure",
  "student": {
  },
  "errors": [
    "name can't be blank"
  ]
}
```

Comment (dearshrewdwit) [reply src/components/Students/StudentInfo.js:28]: Then you can check the status from the response and act accordingly

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]

File: src/components/Students/StudentInfo.js (hunk 2 above)
Comment (RyanVanDijck) [reply src/components/Students/StudentInfo.js]: Changed Here:
https://github.com/digital-futures-academy/atlas-frontend/pull/25/commits/c3a2e3b538bb8bfe9fc6609829ed1b0cebc2f7cf

File: src/components/Students/StudentInfo.js (hunk 3 above)
Comment (RyanVanDijck) [reply src/components/Students/StudentInfo.js]: Changed Here:
https://github.com/digital-futures-academy/atlas-frontend/pull/25/commits/c3a2e3b538bb8bfe9fc6609829ed1b0cebc2f7cf

**Comment by dearshrewdwit** [comment]
These test fail locally for me - can you investigate as well?

**Comment by RyanVanDijck** [comment]
> ?

OK, I'll have a look.

File: cypress/integration/studentInfo.spec.js [hunk 4]
```diff
@@ -1,16 +1,51 @@
...
+        cy.get('#lastName').should('contain', 'Testerson') 
+        cy.get('#github').should('contain', 'testname') 
+        cy.get('#email').should('contain', 'real@email.com') 
+        cy.get('#githubImage').should('exist')
+        cy.get('#response').should('contain', 'success')
+        
+    })
+    it("Failing update", function() {
+        cy.visit('/students/1')
+        cy.intercept("http://localhost:5000/api/students/1", {status: 'failure'})
```
Comment (dearshrewdwit) [inline cypress/integration/studentInfo.spec.js:37]: Check with @bferenczy0 , the API now sends back a consistent response, a failure has an error message you can use

Comment (RyanVanDijck) [reply cypress/integration/studentInfo.spec.js]: Changed here: 
https://github.com/digital-futures-academy/atlas-frontend/pull/25/commits/2c402e5792223a854d4fcb706f7124c6488ae55a

File: src/components/Students/StudentProfile/StudentInfo.js [hunk 5]
```diff
@@ -10,37 +10,101 @@ class StudentInfo extends Component{
...
-        .then(data =>  this.setState({avatar_Url: data.avatar_url}))
+        .then(apiData => {
+          fetch(`https://api.github.com/users/${apiData.student.githubUsername}`)
+            .then(res => res.json())
+            .then(githubData => this.setState({
+              studentData: apiData.student,
+              avatar_Url: githubData.avatar_url
+            }))
+        })
+        // .catch() when a student doesn't exist?
```
Comment (dearshrewdwit) [inline src/components/Students/StudentProfile/StudentInfo.js:30]: Remove these comments by adding an issue to your board with this work. And then we can merge this

**Review by dearshrewdwit — CHANGES_REQUESTED** [review]

File: cypress/integration/studentInfo.spec.js [hunk 6]
```diff
@@ -53,4 +53,26 @@ describe("Student Info", function() {
       cy.get('#response').should('contain', 'No input')
     })
   })
+
+  describe("Failed update for name too long", function() {
```
Comment (dearshrewdwit) [inline cypress/integration/studentInfo.spec.js:57]: @bferenczy0 is this error message that gets returned from the API in this instance?

Comment (dearshrewdwit) [reply cypress/integration/studentInfo.spec.js:57]: >'Data over character limit'

File: cypress/integration/studentInfo.spec.js [hunk 7]
```diff
@@ -53,4 +53,26 @@ describe("Student Info", function() {
...
+    beforeEach(function() {
+      cy.intercept({ url: "http://localhost:5000/api/students/1", method: "PUT" }, { status: 'failure', errors: "Data over character limit" })
+    })
+
+    it("Shows error message for no input", function() {
+      cy.get('#firstName').should('contain', 'Mike')
+      cy.get('#lastNameUpdate').click({force: true})
+      cy.get('#updateInput').type('J'.repeat(301))
```
Comment (dearshrewdwit) [inline cypress/integration/studentInfo.spec.js:65]: ha - nice use of `repeat()`

File: cypress/integration/studentInfo.spec.js (hunk 6 above)
Comment (RyanVanDijck) [reply cypress/integration/studentInfo.spec.js:57]: I don't think this check was implemented on the back end in AT-61

Comment (dearshrewdwit) [reply cypress/integration/studentInfo.spec.js:57]: https://github.com/digital-futures-academy/atlas-api/blob/main/routes/students.js#L50

Comment (dearshrewdwit) [reply cypress/integration/studentInfo.spec.js:57]: Looks like it catches any error from the update (a Sequelize error) and returns a response with the error message

Comment (RyanVanDijck) [reply cypress/integration/studentInfo.spec.js:57]: I'm not quite sure what that error would look like.

Comment (dearshrewdwit) [reply cypress/integration/studentInfo.spec.js:57]:
```sh
curl -X PUT "http://localhost:5000/api/students/1" -d 'firstName=EdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdwardEdward'
{"status":"failure","student":{},"errors":"value too long for type character varying(255)"}
```

Comment (RyanVanDijck) [reply cypress/integration/studentInfo.spec.js:57]: Thank you. 
https://github.com/digital-futures-academy/atlas-frontend/commit/cca39a4ce3b85b83672a2e895f93f2078d777c50

**Review by dearshrewdwit — APPROVED** [review]

**Review by bferenczy0 — APPROVED** [review]

**Pull request merged**